### DISCUSSION TITLE
2021-11-02 weekly planning

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#794] [Maintainer discussions]
### About
Each Tuesday, we start a discussion about Mathesar's progress and general updates. Everyone is welcome to add a new topic to discuss or to comment on existing topics.

### Discussion topics
- [Comms Assignee schedule](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1579172)
- [Milestone: General Improvements 4](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1579198)
- [Product process & Views issues](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1579226)
- [Brent's check-in](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1581967)
- [Sean's check-in](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1582204)
- [Ghislaine's weekly check-in](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1584648)
- [Kriti's check-in](https://github.com/centerofci/mathesar/discussions/794#discussioncomment-1584997)

### Weekly check-in
@mathemancer, @ghislaineguerin, @pavish, @seancolsen, please respond to the weekly check-in as a top-level response and edit the discussion topics above to add your check in as a topic. Others are also welcome to respond.

- What went well over the last week?
- What could have gone better over the last week?
- What are you planning to work on over the next week?
- Do you anticipate any blockers? Can anyone else on the team help?
- Is this weekly check-in helpful to you? How so?
- Anything else you'd like to add?

**Comment by kgodey** [comment]
## Comms Assignee schedule

Upcoming Comms Assignee schedule:
| Week | Person |
|-|-|
| Nov 1 to Nov 5 | Sean |
| Nov 8 to Nov 12 | Ghislaine |
| Nov 15 to Nov 19 | Brent |
| Nov 22 to Nov 26 | Kriti |
| Nov 29 to Dec 3 | Pavish |

You can also find the schedule and instructions for the Comms Assignee on the [wiki page](https://wiki.mathesar.org/team/comms-assignee). Please feel free to update the wiki page if you think of anything useful during your shift.

**Comment by kgodey** [comment]
## Milestone: General Improvements 4 

We only have a few issues left in the next milestone, [General Improvements #4](https://github.com/centerofci/mathesar/milestone/31) and I'd like to close it this week.

- #421 - this issue is on my plate to review and merge
- #635 - this issue only needs a couple of tests to be written and is also on my plate. 
- #639 - @pavish, will you be able to wrap this up this week?
- #651 - @pavish, this is currently assigned to you. Will you be able to complete the URL sync or should @seancolsen take it?
- #748 - @mathemancer, this is in review. Could you make the necessary changes yourself and merge the PR?

**Reply by pavish** [reply]
@kgodey I will be able to complete both #639 and #651 this week.

**Reply by mathemancer** [reply]
I'll look at #748 .  I'm not sure how much still applies with all the recent Dockerfile changes.

**Reply by kgodey** [reply]
Thanks! We should be able to get the milestone completed in that case.

**Comment by kgodey** [comment]
## Product process & Views issues

In order to improve the product process, we're switching to a new issue format (see discussion #763) where design, backend, and frontend work will be tracked in a single issue.

I've started to refactor the issues related to the Views feature to be split into smaller chunks and to explicitly describe user scenarios. As I complete each one, I'm assigning everyone else on the team to review it and then unassign themselves. I'm hoping this will improve our design and implementation process in a few ways:
- Each design spec can cover fewer scenarios and be easier to complete and review.
- Design specs can now include failure / loading / etc. scenarios and hopefully be easier to implement.
- The backend and frontend will have the full context in the same ticket instead of having to read through a bunch of related tickets.

You can see the list of all views issues in this meta issue: #442.

Now that the team has reviewed a couple of tickets, please let me know how it's going and what improvements I can make. One thing that's already on my radar is to describe the high-level purpose of each feature better, perhaps on the wiki.

**Comment by mathemancer** [comment]
> What went well over the last week?

Finally making some progress on the Ranged grouping API work.  

> What could have gone better over the last week?

Still a bit slower velocity on coding than I want.

> * What are you planning to work on over the next week?

I should be able to finish up Ranged Grouping.  The API changes are going to be more minor than I originally thought; they should be additive (I hope) to avoid breaking the current contract.

> * Do you anticipate any blockers? Can anyone else on the team help?

Nothing comes to mind.

> * Is this weekly check-in helpful to you? How so?

The main thing I find helpful is the non-status update parts, i.e., the discussion topics.  I also occasionally notice something out-of-sync in the team members' status updates, but that's more rare.

> * Anything else you'd like to add?

Nothing at the moment.

**Reply by kgodey** [reply]
Thanks! I'm considering shortening the weekly check-in questions next week and making it more free form. Any thoughts about that?

**Comment by seancolsen** [comment]
What went well over the last week?

- Kriti's switch to more "vertical" tickets (with narrow focus, from start to finish) seems good so far.
- Ghislaine's adjustment to solicit feedback earlier in the design process also seems good.
- I felt good about our process of reaching a rough consensus around [NULL vs empty string](https://github.com/centerofci/mathesar/discussions/774) and [NULL terminology](https://github.com/centerofci/mathesar/discussions/778).

What could have gone better over the last week?

- I've been feeling a bit overwhelmed with the amount of preliminary work that seems to be necessary before I can start coding. I'm used to beginning work on a ticket which is much more clearly defined, within a team where more decisions have already been formed, and working with code that doesn't require as much refactoring. I briefly chatted about these challenges with Kriti during our check-in yesterday and she encouraged me to embrace the need to spend time on this preliminary work because it's important, given the relative immaturity of our product.
- I'd like to push myself towards more quickly accepting other people's opinions and design choices, even when I hold differing opinions. I'm used to either: (a) working on a very well-established product which sets clear precedents for design choices going forward, or (b) making design choices in one-on-one synchronous conversations with freelance clients, or (c) building a product that I don't plan to use (and thus have little opinion about). This new role is unique for me -- and exciting -- because I'm working asynchronously with a team in the very early stages of building a product that I actually want to use! I've been thinking for years about how I would design a product like Mathesar, so admittedly I've arrived with some pre-conceived notions. I'd like to carry some of those ideas into my work here but let go of any attachment I have to particular outcomes.
- I'm still at a point where I'm getting lots of merge conflicts with PRs. My approach to PRs is to try making them as small as possible. But it seems like perhaps the culture here is leaning towards larger PRs. I'm still getting a feel for it. But it's been a little tough finding the right balance. Going forward, I plan to try opening more PRs in sequence rather than in parallel (e.g. PR 2 will depend on PR 1).

What are you planning to work on over the next week?

- Still chipping away at the pieces of [setting a column as non-nullable or unique](https://github.com/centerofci/mathesar/issues/499).

Do you anticipate any blockers? Can anyone else on the team help?

- Nope
  

Anything else you'd like to add?

- I spent a little time playing with NocoDB to better understand how they do things. From skimming their promo materials I had high expectations. But after playing with the actual app, I was somewhat surprised to find widespread bugs, poor UX, and very limited features. Overall, very underwhelming. It will be interesting to keep an eye on their development though. I was impressed with the ease of their "deploy to Heroku" setup process, not having used Heroku before. We'll need to consider options like that at some point.

**Reply by seancolsen** [reply]
Oops, I had a template from last week that I've been filling in as I go, so I missed this new question:

> Is this weekly check-in helpful to you? How so?

I'm not sure. I'm still finding my groove here. This check-in format is new to me. In some ways it's helpful to have the continued reflection. In other ways it's a distraction from more meaningful work. I haven't had enough time with the process to form a solid opinion yet.

**Reply by kgodey** [reply]
As we discussed yesterday, I'm hoping that we will get better at reducing preliminary work as the product matures. I'm open to suggestions about how to improve the process, so let me know if you have any thoughts about that in the future.

> I'm still at a point where I'm getting lots of merge conflicts with PRs. 

We also talked about this a bit today. I think we need to coordinate frontend features between you and @pavish better so that you're not working on the same areas of the codebase. I think that after you finish your in-progress work, it might make sense for @pavish to finish up the "Working with Tables" milestone (including the open PRs from contributors) and you to work on something independent like the toast system.

@pavish I think you and Sean should talk about this issue too and follow up here after you've had a chance to discuss.

**Reply by pavish** [reply]
@seancolsen @kgodey I would like to co-ordinate better to reduce merge conflicts. However, I do think that we can only do that to a small extent. Unlike the backend where each feature/api have clear separations, the frontend uses the same visual components and work on the same area of the UI, even though each issue covers different features.

For eg., the issues in "Working with Tables" milestone all operate on the actions pane and on the column dropdown. Same applies for all the milestones that come after this. There is no possible way to avoid merge conflicts here, unless one of us is working on a completely different milestone/independent feature, even then we'd have merge conflicts with PRs from community contributors.

I think we should get used to this, expect this to happen, and even accept this as a necessary step to completing milestones. I would like to find ways to make it easier for us though.

**Comment by ghislaineguerin** [comment]
> What went well over the last week?

Design discussions are very focused and helpful for me. I feel that we are all in sync regarding product vision. 

> What could have gone better over the last week?

Better documentation for design could resolve some of Sean's questions as he tries to catch up with the decisions we've been making so far.

> What are you planning to work on over the next week?

Start prototyping for views.

> Do you anticipate any blockers? Can anyone else on the team help?

None

> Is this weekly check-in helpful to you? How so?

It is. I like reflecting on my week in general.

> Anything else you'd like to add?

The HTML prototype makes me feel more efficient and focused, but at the same time might not be accessible for others. I need to figure this out, but for now, I'd like to keep prototyping w code.

**Reply by kgodey** [reply]
> The HTML prototype makes me feel more efficient and focused, but at the same time might not be accessible for others. I need to figure this out, but for now, I'd like to keep prototyping w code.

I haven't noticed any problems with accessibility. Are there specific problems with the prototype that you're thinking of?

**Comment by kgodey** [comment]
> What went well over the last week?

I was finally able to get into the Views issues. I'm also glad interviews have been completed.

> What could have gone better over the last week?

I would have liked to get more done, but I'm not sure there's anything I could have done any better.

> What are you planning to work on over the next week?

- Speccing out the remaining Views issues
- Finalizing hiring and the 2022 budget
- Merging the two open PRs for the next milestone
- Moving design and process issues forward
- Defining common error structure for API

> Is this weekly check-in helpful to you? How so?

Yes, it's helpful to see what everyone is working on. It takes some of my time to set up, though and I'm not sure if it's worth doing it. There's rarely much discussion outside of people answering check in questions, so that's why I'm asking this.

> - Do you anticipate any blockers? Can anyone else on the team help?
> - Anything else you'd like to add?

No.

**Comment by pavish** [comment]
> What went well over the last week?

The column type PR is close to completion and the approach for the E2E test environment is all figured out, it will be complete soon. I'm also glad that our design discussions around links have been very focused.

> What could have gone better over the last week?

I reviewed most PRs at the start of the week, and then tried to take some focused time for coding. This lead to a set of new PRs getting queued up. A switch in context slows me down, I still haven't figured out a way to balance that between reviewing and writing code. 

> What are you planning to work on over the next week?

* Get [Change column type PR](https://github.com/centerofci/mathesar/pull/786) merged
* Get [E2E integ test environment PR](https://github.com/centerofci/mathesar/pull/746) merged
* Complete url sync changes which was delayed from the issue: https://github.com/centerofci/mathesar/issues/651
* Review and merge all current open frontend PRs
* Work on setting a default value for column, if I get time for it

> Do you anticipate any blockers? Can anyone else on the team help?

None

> Is this weekly check-in helpful to you? How so?

Yes, I like a considated view of what I've done in the week. When I self-estimate the time to complete issues, I tend to consider little time towards non-coding tasks, leading to overestimating my capacity towards milestone completion. A weekly check-in keeps me in place and helps me take everything into account.

> Anything else you'd like to add?

Nothing at the moment.
